src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1450 @@
+        indices: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        seq_len: Optional[int] = None,
+    ) -> Union[Tuple[torch.Tensor, ...], BaseModelOutput]:
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if batch_size is None and seq_len is None:
+            batch_size, seq_len = input_ids.shape[:2]
+
+        if attention_mask is None:
+            attention_mask = torch.ones((batch_size, seq_len), device=input_ids.device, dtype=torch.bool)
+
+        repad = False
+        if self.config.unpad_inputs:
+            if indices is None and cu_seqlens is None and max_seqlen is None:
+                repad = True
+                if self.config.unpad_no_grad:
+                    input_ids, indices, cu_seqlens, max_seqlen, *_ = self._unpad_inputs_no_grad(
+                        input_ids, attention_mask
+                    )
+                else:
+                    input_ids, indices, cu_seqlens, max_seqlen, *_ = self._unpad_inputs(input_ids, attention_mask)
+        elif position_ids is None:
+            position_ids = torch.arange(seq_len, device=input_ids.device).unsqueeze(0)
+
+        hidden_states = self.embeddings(input_ids)
+
+        # expand attention_mask
+        if self.config._attn_implementation != "flash_attention_2" and attention_mask is not None:
+            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+            attention_mask = _prepare_4d_attention_mask(attention_mask, hidden_states.dtype)
+
+        for encoder_layer in self.layers:
+            if self.gradient_checkpointing and self.training:
+                hidden_states = self._gradient_checkpointing_func(
+                    encoder_layer.__call__,
+                    hidden_states,
+                    attention_mask,
+                    position_ids,
+                    cu_seqlens,
+                    max_seqlen,
+                )
+            else:
+                hidden_states = encoder_layer(
+                    hidden_states,
+                    attention_mask=attention_mask,
+                    position_ids=position_ids,
+                    cu_seqlens=cu_seqlens,
+                    max_seqlen=max_seqlen,
+                )
+
+        hidden_states = self.final_norm(hidden_states)
+
+        if repad:
+            hidden_states, _ = self._pad_outputs(hidden_states, indices, batch_size, seq_len)
+
+        if not return_dict:
+            return hidden_states
+        return BaseModelOutput(last_hidden_state=hidden_states)
+
+
+class ModernBertForMaskedLM(ModernBertPreTrainedModel):
+    def __init__(
+        self,
+        config: ModernBertConfig,
+        pad_logits: bool = True,
+        pad_logits_no_grad: Optional[bool] = None,
+        sparse_prediction: bool = True,
+        sparse_pred_ignore_index: int = -100,
```

Comment:
Let's try and get rid of these. Either via the `config`, or avoid needing them altogether.
I also feel like `pad_logits` should be inferrable via `config.unpad_inputs` and `pad_logits_no_grad` via `config.unpad_no_grad`.

I'm not sure what we should do with the `sparse_prediction` stuff.

Reply:
Resolved via 38f959bf096bd05e1174ed4495daac004bf115fa, always open to suggestions